tools/bpf/bpftool/gen.c

```diff
@@ -1097,9 +1100,887 @@ static int do_help(int argc, char **argv)
+			}
+			break;
+		case BTF_KIND_PTR:
+		case BTF_KIND_TYPEDEF:
+		case BTF_KIND_VOLATILE:
+		case BTF_KIND_CONST:
+		case BTF_KIND_RESTRICT:
+		case BTF_KIND_FUNC:
+		case BTF_KIND_VAR:
+			btf_type->type = btf_reloc_id_get(info, btf_type->type);
+			break;
+		case BTF_KIND_ARRAY:
+			array = btf_array(btf_type);
+			array->index_type = btf_reloc_id_get(info, array->index_type);
+			array->type = btf_reloc_id_get(info, array->type);
+			break;
+		case BTF_KIND_FUNC_PROTO:
+			btf_type->type = btf_reloc_id_get(info, btf_type->type);
+			params = btf_params(btf_type);
+			for (unsigned short j = 0; j < btf_vlen(btf_type); j++)
+				params[j].type = btf_reloc_id_get(info, params[j].type);
+			break;
+		default:
+			break;
+		}
+	}
+
+	return btf_new;
+
+out:
+	btf__free(btf_new);
+	return ERR_PTR(err);
+}
+
+#define MAX_OBJECTS 128
+
+static int is_file(const char *path)
+{
+	struct stat st;
+
+	if (stat(path, &st) < 0)
+		return -1;
+
+	switch (st.st_mode & S_IFMT) {
+	case S_IFDIR:
+		return 0;
+	case S_IFREG:
+		return 1;
+	default:
+		return -1;
+	}
+
+	return -1;
+}
+
+static int generate_btf(const char *src_btf, const char *dst_btf, const char *objspaths[])
+{
+	struct btf_reloc_info *reloc_info;
+	struct bpf_object *obj;
+	struct btf *btf_new = NULL;
+	int err;
+
+	struct bpf_object_open_opts ops = {
+		.sz = sizeof(ops),
+		.btf_custom_path = src_btf,
+	};
+
+	reloc_info = btfgen_reloc_info_new(src_btf);
+	err = libbpf_get_error(reloc_info);
+	if (err) {
+		p_err("failed to allocate info structure");
+		goto out;
+	}
+
+	for (int i = 0; objspaths[i] != NULL; i++) {
+		printf("OBJ : %s\n", objspaths[i]);
+		obj = bpf_object__open_file(objspaths[i], &ops);
+		err = libbpf_get_error(obj);
+		if (err) {
+			p_err("error opening object: %s", strerror(errno));
+			goto out;
+		}
+
+		err = btfgen_obj_reloc_info_gen(reloc_info, obj);
+		if (err)
+			goto out;
+
+		bpf_object__close(obj);
+	}
+
+	btf_new = btfgen_reloc_info_get_btf(reloc_info);
+	err = libbpf_get_error(btf_new);
+	if (err) {
+		p_err("error generating btf: %s", strerror(errno));
+		goto out;
+	}
+
+	printf("DBTF: %s\n", dst_btf);
+	err = btf_save_raw(btf_new, dst_btf);
+	if (err) {
+		p_err("error saving btf file: %s", strerror(errno));
+		goto out;
+	}
+
+out:
+	if (!libbpf_get_error(btf_new))
+		btf__free(btf_new);
+	btfgen_reloc_info_free(reloc_info);
+
+	return err;
+}
+
+static int do_gen_btf(int argc, char **argv)
+{
+	char src_btf_path[PATH_MAX], dst_btf_path[PATH_MAX];
+	bool input_is_file, output_is_file;
+	const char *objs[MAX_OBJECTS];
+	const char *input, *output;
+	struct dirent *dir;
+	int i, err;
+	DIR *d;
+
+	if (!REQ_ARGS(3)) {
+		usage();
+		return -1;
+	}
+
+	input = GET_ARG();
+	err = is_file(input);
+	if (err < 0) {
+		p_err("failed to stat %s: %s", input, strerror(errno));
+		return err;
+	}
+	input_is_file = err;
+
+	output = GET_ARG();
+	err = is_file(output);
+	if (err < 0) {
+		p_err("failed to stat %s: %s", output, strerror(errno));
```

Comment:
This prevents the output file from being created, we probably want to avoid giving an error here and just create the file if the path is not a folder

Reply:
Yup, maybe it is worth opening the file for writing ("w") which creates the file if not existing already and truncates it if already existing (if I do not remember it wrong).

Reply:
Good catch! The file is already opened in "wb" mode, but this check was incorrect. I changed it to assume that if the file doesn't exist then the output mode is "file".